Intern/rayx-ui/src/Triangulation/Triangulate.cpp

```diff
@@ -2,128 +2,423 @@
+                    stack.push_back(last);
+                    stack.push_back(current);
+                } else {
+                    uint32_t currentInside = current.first;
+                    auto beforeCurrent = last;
+                    bool firstIteration = true;
+                    while (stack.size() > 0) {
+                        auto diag = split(currentInside, last.first);
+                        last = stack.back();
+                        stack.pop_back();
+                        currentInside = diag.first;
+                        if (firstIteration) {
+                            firstIteration = false;
+                            beforeCurrent.first = diag.second;
+                        }
+                    }
+                    stack.push_back(beforeCurrent);
+                    stack.push_back(current);
+                }
+            } else {
+                if (last.second == Right) {
+                    while (!stack.empty() && checkAngle(Right, current.first, last.first, stack.back().first)) {
+                        auto diag = split(current.first, stack.back().first);
+                        last = stack.back();
+                        stack.pop_back();
+                        current.first = diag.first;
+                    }
+                    stack.push_back(last);
+                    stack.push_back(current);
+                } else {
+                    auto currentOutside = current;
+                    auto lastCopy = last;
+                    bool firstIteration = true;
+                    while (stack.size() > 0) {
+                        auto diag = split(current.first, last.first);
+                        last = stack.back();
+                        stack.pop_back();
+                        if (firstIteration) {
+                            firstIteration = false;
+                            currentOutside.first = diag.first;
+                        }
+                    }
+                    stack.push_back(lastCopy);
+                    stack.push_back(currentOutside);
+                }
+            }
+        }
+        auto bottom = sortedSequence.back();
+        stack.pop_back();
+        auto last = stack.back();
+        stack.pop_back();
+        if (last.second == Left) {
+            while (!stack.empty()) {
+                split(bottom.first, last.first);
+                last = stack.back();
+                stack.pop_back();
+            }
+        } else {
+            while (!stack.empty()) {
+                auto diag = split(bottom.first, last.first);
+                bottom.first = diag.first;
+                last = stack.back();
+                stack.pop_back();
+            }
+        }
+    }
+
+    bool isTop(uint32_t edge) { return getOrigin(getPrev(edge)) < getOrigin(edge) && getOrigin(getNext(edge)) < getOrigin(edge); }
+};
+
+void triangulate(PolygonComplex poly, std::vector<TextureVertex>& points, std::vector<uint32_t>& indices) {
```

Comment:
Parameters that are not trivial data types (int, char, double) should be passed as reference or const ref. Especially when the function is called often